File: src/main/java/translating/StatementLoader.java
package translating;

import facts.parameters.InParameter;
import facts.parameters.OutParameter;
import facts.exp.*;
import facts.stmt.*;
import facts.stmt.Assign.Load;
import facts.stmt.Assign.Move;
import facts.stmt.Assign.Store;
import org.antlr.v4.runtime.ParserRuleContext;
import org.antlr.v4.runtime.tree.ErrorNode;
import org.antlr.v4.runtime.tree.TerminalNode;
import java.util.List;
import BilParser.*;


public class StatementLoader implements BilListener {

    // list of facts to output
    List<Stmt> stmts;
    // the last function header parsed; needed for assigning facts.parameters
    EnterSub currentFunction;
    // for generating unique labels
    int pcCount = 0;

    public StatementLoader(List<Stmt> stmts) {
        this.stmts = stmts;
    }

    private Expr parseExpression(BilParser.ExpContext ctx) {
        if (ctx == null) {
            return null;
        } else if (ctx.getClass().equals(BilParser.ExpBopContext.class)) {
            return parseBinaryOperation((BilParser.ExpBopContext) ctx);
        } else if (ctx.getClass().equals(BilParser.ExpUopContext.class)) {
            return parseUnaryOperation((BilParser.ExpUopContext) ctx);
        } else if (ctx.getClass().equals(BilParser.ExpVarContext.class)) {
            return parseVariableExpression((BilParser.ExpVarContext) ctx);
        } else if (ctx.getClass().equals(BilParser.ExpLiteralContext.class)) {
            return parseLiteralExpression((BilParser.ExpLiteralContext) ctx);
        } else if (ctx.getClass().equals(BilParser.ExpExtractContext.class)) {
            return parseExtractionExpression((BilParser.ExpExtractContext) ctx);
        } else if (ctx.getClass().equals(BilParser.ExpCastContext.class)) {
            return parseCastExpression((BilParser.ExpCastContext) ctx);
        } else if (ctx.getClass().equals(BilParser.ExpLoadContext.class)) {
            // toReturn == null when this load is not a mem expression, which is currently not expected
            Expr toReturn = parseLoadExpression((BilParser.ExpLoadContext) ctx);
            if (toReturn != null) {
                return toReturn;
            }
        } else if (ctx.getClass().equals(BilParser.ExpBracketContext.class)) {
            return parseBracketExpression((BilParser.ExpBracketContext) ctx);
        }
        System.err.println("Unhandled expression detected: " + ctx.getText());
        return null;
    }

    private Expr parseBracketExpression(BilParser.ExpBracketContext ctx) {
        return parseExpression(ctx.exp());
    }

    private BinOp parseBinaryOperation(BilParser.ExpBopContext ctx) {
        Expr left = parseExpression(ctx.exp(0));
        Expr right = parseExpression(ctx.exp(1));
        String op = ctx.bop().getText();
        if (op.equals("=")) {
            op = "=="; // a little patching here, a little there...
        }
        return new BinOp(op, left, right);
    }

    private UniOp parseUnaryOperation(BilParser.ExpUopContext ctx) {
        Expr exp = parseExpression(ctx.exp());
        String op = ctx.uop().getText();
        return new UniOp(op, exp);
    }

    private Var parseVariableExpression(BilParser.ExpVarContext ctx) {
        return new Var(ctx.var().getText());
    }

    private Literal parseLiteralExpression(BilParser.ExpLiteralContext ctx) {
        return new Literal(ctx.literal().getText());
    }

    // fixme: assumes all bit vectors are 64 bits long
    private Extract parseExtractionExpression(BilParser.ExpExtractContext ctx) {
        int firstNat = 64 - Integer.parseInt(ctx.nat(0).getText());
        int secondNat = 63 - Integer.parseInt(ctx.nat(1).getText());
        Expr exp = parseExpression(ctx.exp());
        return new Extract(firstNat, secondNat, exp);
    }

    private Expr parseCastExpression(BilParser.ExpCastContext ctx) {
        // simply unwrap and throw away casts
        return parseExpression(ctx.exp());
    }

    private MemExpr parseLoadExpression(BilParser.ExpLoadContext ctx) {
        if (ctx.exp(0).getText().equals("mem")) {
            return new MemExpr(parseExpression(ctx.exp(1)));
        } else {
            return null;
        }
    }

    private String uniquePc() {
        int pc = pcCount++;
        return "l" + pc;
    }

    @Override
    public void enterBil(BilParser.BilContext ctx) {

    }

    @Override
    public void exitBil(BilParser.BilContext ctx) {

    }

    @Override
    public void enterFunction(BilParser.FunctionContext ctx) {

    }

    @Override
    public void exitFunction(BilParser.FunctionContext ctx) {

    }

    @Override
    public void enterProgdecl(BilParser.ProgdeclContext ctx) {

    }

    @Override
    public void exitProgdecl(BilParser.ProgdeclContext ctx) {

    }

    @Override
    public void enterSub(BilParser.SubContext ctx) {
        String address = ctx.addr().getText();
        String name = ctx.functionName().getText();
        EnterSub function = new EnterSub(address, name);
        stmts.add(function);

        this.currentFunction = function;
    }

    @Override
    public void exitSub(BilParser.SubContext ctx) {

    }

    @Override
    public void enterParamTypes(BilParser.ParamTypesContext ctx) {
        String id = ctx.param().getText(); // human-readable name
        String variable = ctx.var().getText(); // some register, probably
        if (id.contains("result")) {
            currentFunction.setOutParam(new OutParameter(new Var("out"), new Var(variable)));
        } else {
            currentFunction.getInParams().add(new InParameter(new Var(id), new Var(variable)));
        }
    }

    @Override
    public void exitParamTypes(BilParser.ParamTypesContext ctx) {

    }

    @Override
    public void enterStmt(BilParser.StmtContext ctx) {
        String address = ctx.addr().getText();
        // statements can be assignments, jumps, conditional jumps or function calls
        if (ctx.assign() != null) {
            // statement is assignment; check which type
            BilParser.AssignContext assignCtx = ctx.assign();
            if (assignCtx.exp().getClass().equals(BilParser.ExpLoadContext.class)) {
                // statement is a load assignment
                BilParser.ExpLoadContext loadCtx = (BilParser.ExpLoadContext) assignCtx.exp();
                Var lhs = new Var(loadCtx.exp(1).getText());
                Expr rhs = parseExpression(loadCtx.exp(2));
                if (rhs != null) { // null check is necessary as rhs may not exist for loads
                    stmts.add(new Load(address, lhs, (MemExpr) rhs));
                }
            } else if (assignCtx.exp().getClass().equals(BilParser.ExpStoreContext.class)) {
                // statement is a store assignment
                BilParser.ExpStoreContext storeCtx = (BilParser.ExpStoreContext) assignCtx.exp();
                MemExpr lhs = new MemExpr(parseExpression(storeCtx.exp(1)));
                Expr rhs = parseExpression(storeCtx.exp(2));
                stmts.add(new Store(address, lhs, rhs));
            } else {
                // statement is a move assignment
                Var lhs = new Var(assignCtx.var().getText());
                Expr rhs = parseExpression(assignCtx.exp());
                stmts.add(new Move(address, lhs, rhs));
            }
        } else if (ctx.jmp() != null) {
            // statement is a jump
            String target = "";
            if (ctx.jmp().var() != null) {
                target = ctx.jmp().var().getText();
            } else if (ctx.jmp().addr() != null) {
                target = ctx.jmp().addr().getText();
            }
            stmts.add(new JmpStmt(address, target));
        } else if (ctx.cjmp() != null) {
            // statement is a conditional jump
            Var cond = new Var(ctx.cjmp().var().getText()); // conditions are always vars
            String target = ctx.cjmp().addr().getText();
            stmts.add(new CJmpStmt(address, target, cond));
        } else if (ctx.call() != null) {
            // statement is a call
            if (ctx.call().functionName() == null) {
                // occasionally this occurs with "call LR with no return" lines
                stmts.add(new ExitSub(ctx.addr().getText()));
            } else {
                String funcName = ctx.call().functionName().getText();
                String returnAddr = null;
                if (ctx.call().returnaddr().addr() != null) {
                    returnAddr = ctx.call().returnaddr().addr().getText();
                }
                stmts.add(new CallStmt(address, funcName));

                // handle the case of no return
                if (returnAddr != null) stmts.add(new JmpStmt(uniquePc(), returnAddr));
            }
        } else {
            // this statement is empty
            stmts.add(new SkipStmt(address));
        }
    }

    @Override
    public void exitStmt(BilParser.StmtContext ctx) {

    }

    @Override
    public void enterCall(BilParser.CallContext ctx) {
    }

    @Override
    public void exitCall(BilParser.CallContext ctx) {

    }

    @Override
    public void enterAssign(BilParser.AssignContext assignCtx) {

    }

    @Override
    public void exitAssign(BilParser.AssignContext ctx) {

    }

    @Override
    public void enterExpBracket(BilParser.ExpBracketContext ctx) {

    }

    @Override
    public void exitExpBracket(BilParser.ExpBracketContext ctx) {

    }

    @Override
    public void enterExpUop(BilParser.ExpUopContext ctx) {

    }

    @Override
    public void exitExpUop(BilParser.ExpUopContext ctx) {

    }

    @Override
    public void enterExpVar(BilParser.ExpVarContext ctx) {

    }

    @Override
    public void exitExpVar(BilParser.ExpVarContext ctx) {

    }

    @Override
    public void enterExpLiteral(BilParser.ExpLiteralContext ctx) {

    }

    @Override
    public void exitExpLiteral(BilParser.ExpLiteralContext ctx) {

    }

    @Override
    public void enterExpExtract(BilParser.ExpExtractContext ctx) {

    }

    @Override
    public void exitExpExtract(BilParser.ExpExtractContext ctx) {

    }

    @Override
    public void enterExpCast(BilParser.ExpCastContext ctx) {

    }

    @Override
    public void exitExpCast(BilParser.ExpCastContext ctx) {

    }

    @Override
    public void enterExpLoad(BilParser.ExpLoadContext ctx) {

    }

    @Override
    public void exitExpLoad(BilParser.ExpLoadContext ctx) {

    }

    @Override
    public void enterExpStore(BilParser.ExpStoreContext ctx) {

    }

    @Override
    public void exitExpStore(BilParser.ExpStoreContext ctx) {

    }

    @Override
    public void enterExpBop(BilParser.ExpBopContext ctx) {

    }

    @Override
    public void exitExpBop(BilParser.ExpBopContext ctx) {

    }

    @Override
    public void enterCjmp(BilParser.CjmpContext ctx) {

    }

    @Override
    public void exitCjmp(BilParser.CjmpContext ctx) {

    }

    @Override
    public void enterJmp(BilParser.JmpContext ctx) {

    }

    @Override
    public void exitJmp(BilParser.JmpContext ctx) {

    }

    @Override
    public void enterVar(BilParser.VarContext ctx) {

    }

    @Override
    public void exitVar(BilParser.VarContext ctx) {

    }

    @Override
    public void enterFunctionName(BilParser.FunctionNameContext ctx) {

    }

    @Override
    public void exitFunctionName(BilParser.FunctionNameContext ctx) {

    }

    @Override
    public void enterParam(BilParser.ParamContext ctx) {

    }

    @Override
    public void exitParam(BilParser.ParamContext ctx) {

    }

    @Override
    public void enterBop(BilParser.BopContext ctx) {

    }

    @Override
    public void exitBop(BilParser.BopContext ctx) {

    }

    @Override
    public void enterUop(BilParser.UopContext ctx) {

    }

    @Override
    public void exitUop(BilParser.UopContext ctx) {

    }

    @Override
    public void enterInout(BilParser.InoutContext ctx) {

    }

    @Override
    public void exitInout(BilParser.InoutContext ctx) {

    }

    @Override
    public void enterReturnaddr(BilParser.ReturnaddrContext ctx) {

    }

    @Override
    public void exitReturnaddr(BilParser.ReturnaddrContext ctx) {

    }

    @Override
    public void enterLiteral(BilParser.LiteralContext ctx) {

    }

    @Override
    public void exitLiteral(BilParser.LiteralContext ctx) {

    }

    @Override
    public void enterNat(BilParser.NatContext ctx) {

    }

    @Override
    public void exitNat(BilParser.NatContext ctx) {

    }

    @Override
    public void enterAddr(BilParser.AddrContext ctx) {

    }

    @Override
    public void exitAddr(BilParser.AddrContext ctx) {

    }

    @Override
    public void visitTerminal(TerminalNode terminalNode) {

    }

    @Override
    public void visitErrorNode(ErrorNode errorNode) {

    }

    @Override
    public void enterEveryRule(ParserRuleContext parserRuleContext) {

    }

    @Override
    public void exitEveryRule(ParserRuleContext parserRuleContext) {

    }
}

File: src/main/scala/translating/StatementLoader.scala
package translating

import astnodes.*
import BilParser.BilParser.*
import BilParser.*
import org.antlr.v4.runtime.ParserRuleContext
import org.antlr.v4.runtime.tree.{ErrorNode, ParseTree, ParseTreeProperty, TerminalNode}
import FlowGraph.Function
import vcgen.State
import util.AssumptionViolationException

import scala.collection.mutable
import scala.collection.mutable.ArrayBuffer
import scala.jdk.CollectionConverters.*

// TODO create a statementloaderstate class
class StatementLoader() extends BilBaseListener {
  val stmts: ArrayBuffer[Stmt] = ArrayBuffer[Stmt]()
  private val exprs = ParseTreeProperty[Expr]
  private def getExpr(node: ParseTree) =
    if (node == null) throw new NullPointerException("Null expression")
    if (exprs.get(node) != null)
      exprs.get(node)
    else
      throw new Exception("Unparsed expression " + node.getText + " (" + node.getClass + ")")

  private val preds = ParseTreeProperty[Pred]
  private def getPred(node: ParseTree) =
    if (node == null) throw new NullPointerException("Null pred")
    if (preds.get(node) != null) preds.get(node)
    else throw new Exception("Unparsed pred " + node.getText + " (" + node.getClass + ")")


  private val secs = ParseTreeProperty[Sec]
  private def getSec(node: ParseTree) =
    if (node == null) throw new NullPointerException("Null security")
    if (secs.get(node) != null) secs.get(node)
    else throw new Exception("Unparsed security " + node.getText + " (" + node.getClass + ")")

  val lPreds = new mutable.HashMap[Register, Sec]
  val gammaMappings = new mutable.HashMap[Register, SecVar]


  val varSizes: mutable.Map[String, Int] = mutable.Map[String, Int]()

  var rely: Option[Pred] = None

  /*
  private def getVarSize(name: String, rhs: Option[Expr]): Int = {
    if (!varSizes.contains(name)) {
      if (rhs.isEmpty) throw new AssumptionViolationException("")
      if (name.charAt(0) == 'R') varSizes(name) = 64
      else varSizes(name) = rhs.get.size.get
    }
    varSizes(name)
  }
  */

  /**
   * TODO: fix samples so 'X' is unnecessary; confirm pointer sizes for SP, FP, & LR
   */
  private def setVarSize(name: String, rhs: Expr): Unit = {
    if (!varSizes.contains(name)) {
      if (name.charAt(0) == 'R' || name.charAt(0) == 'X') varSizes(name) = 64
      else if (name.charAt(0) == '#') varSizes(name) = rhs.size.get
      else if (name == "NF" || name == "ZF" || name == "CF" || name == "VF") varSizes(name) = 1
      else if (name == "SP" || name == "FP" || name == "LR") varSizes(name) = 64
      else {
        println(name)
        throw new RuntimeException
      }
    }
  }


  // the last function header parsed; needed for assigning parameters
  private var currentFunction: Option[EnterSub] = None

  // for generating unique labels
  private var pcCount = 0

  def typeToSize(typeStr: String): Int = typeStr match {
    case "u32" => 32
    case "u64" => 64
    case _ => ???
  }

  var requires: List[Pred] = List()
  var ensures: List[Pred] = List()

  override def exitSub(ctx: BilParser.SubContext): Unit = {
    currentFunction match {
      case Some(f) =>
        f.requires = requires
        f.ensures = ensures
      case None =>
    }

    requires = List()
    ensures = List()

    val address = ctx.addr.getText
    val name = ctx.functionName.getText

    val function = EnterSub(address, name, List(), List())
    stmts += function
    this.currentFunction = Some(function)
  }
  
  override def exitParamTypes(ctx: BilParser.ParamTypesContext): Unit = {
    val id = ctx.param.getText // human-readable name
    val variable = ctx.`var`.getText // some register, probably

    val size = typeToSize(ctx.nat.getText)

    if (id.contains("result")) {
      currentFunction.get.outParam = Some(OutParameter(Register(id, size), Register(variable, 64)))
    } else {
      currentFunction.get.inParams += InParameter(Register(id, size), Register(variable, size))
    }
  }

  /*
  TODO!!!!
  override def exitEnsuresSpec(ctx: BilParser.EnsuresSpecContext): Unit = ensures = ensures.appended(getPred(ctx.pred))
  override def exitRequiresSpec(ctx: BilParser.RequiresSpecContext): Unit = requires = requires.appended(getPred(ctx.pred))
  */

  override def exitStmt(ctx: BilParser.StmtContext): Unit = {
    val address = ctx.addr.getText
// statements can be assignments, jumps, conditional jumps or function calls
    if (ctx.assign != null) { // statement is assignment; check which type
      val assignCtx = ctx.assign
      val LHS = getExpr(assignCtx.`var`)
      val RHS = getExpr(assignCtx.exp)

      stmts += ((LHS, RHS) match {
        case (v: Register, m: MemStore) =>
          if (v.name == "mem") MemAssign(address, MemLoad(m.loc, m.size), m.expr)
          else throw new AssumptionViolationException("expected mem for memstore")
        case (v: Register, _) =>
          setVarSize(v.name, RHS)
          RegisterAssign(address, v.copy(size = Some(varSizes(v.name))), RHS)
        case _ => throw new AssumptionViolationException("Unexpected expression")
      })

      /*
      if (assignCtx.exp.getClass == classOf[BilParser.ExpLoadContext]) { // statement is a load assignment
        val loadCtx = assignCtx.exp.asInstanceOf[BilParser.ExpLoadContext]
        val lhs = new Var(loadCtx.exp(0).getText)
        val rhs = getExpr(loadCtx.exp(1))
        if (rhs != null) { // null check is necessary as rhs may not exist for loads
          stmts += new Load(address, lhs, rhs.asInstanceOf[MemExpr])
        }
      } else if (assignCtx.exp.getClass == classOf[BilParser.ExpStoreContext]) { // statement is a store assignment
        val storeCtx = assignCtx.exp.asInstanceOf[BilParser.ExpStoreContext]
        val lhs = new MemExpr(getExpr(storeCtx.exp(0)))
        val rhs = getExpr(storeCtx.exp(1))
        stmts += new Store(address, lhs, rhs)
      } else { // statement is a move assignment
        val lhs = new Var(assignCtx.`var`.getText)
        val rhs = getExpr(assignCtx.exp)
        stmts += new Move(address, lhs, rhs)
      }
      */
    } else if (ctx.jmp != null) { // statement is a jump
      var target = ""
      if (ctx.jmp.`var` != null) target = ctx.jmp.`var`.getText
      else if (ctx.jmp.addr != null) target = ctx.jmp.addr.getText
      stmts += JmpStmt(address, target)
    } else if (ctx.cjmp != null) { // statement is a conditional jump
      val cond = Register(ctx.cjmp.`var`.getText, varSizes(ctx.cjmp.`var`.getText)) // conditions are always vars
      val target = ctx.cjmp.addr.getText
      stmts += CJmpStmt(address, target, "TODO", cond)
    } else if (ctx.call != null) { // statement is a call
      if (ctx.call.functionName == null) { // occasionally this occurs with "call LR with no return" lines
        stmts += ExitSub(ctx.addr.getText, None)
      } else {
        val funcName = ctx.call.functionName.getText
        stmts += CallStmt(address, funcName, Option(ctx.call.returnaddr.addr).map(_.getText), List(), None)
      }
    } else { // this statement is empty
      stmts += SkipStmt(address)
    }
  }

  override def exitExpBracket(ctx: BilParser.ExpBracketContext): Unit = exprs.put(ctx, getExpr(ctx.exp))
  
  override def exitExpUop(ctx: BilParser.ExpUopContext): Unit =
    exprs.put(ctx, UniOp(UniOperator.fromBil(ctx.uop.getText), getExpr(ctx.exp)))
    
  override def exitExpBop(ctx: BilParser.ExpBopContext): Unit =
    exprs.put(ctx, BinOp(BinOperator.fromBil(ctx.bop.getText), getExpr(ctx.exp(0)), getExpr(ctx.exp(1))))
    
  override def exitVar(ctx: BilParser.VarContext): Unit =
    exprs.put(ctx, new Register(ctx.getText, varSizes.get(ctx.getText)))
    
  override def exitExpVar(ctx: BilParser.ExpVarContext): Unit = exprs.put(ctx, exprs.get(ctx.`var`))
  
  override def exitExpLiteral(ctx: BilParser.ExpLiteralContext): Unit = exprs.put(ctx, Literal(ctx.literal.getText))
  
  override def exitExpExtract(ctx: BilParser.ExpExtractContext): Unit = {
    val firstNat = ctx.nat(0).getText.toInt
    val secondNat = ctx.nat(1).getText.toInt
    val exp = getExpr(ctx.exp)
    exprs.put(ctx, Extract(firstNat, secondNat, exp))
  }

  // Currently doesn't handle high/low. these could be converted to extracts
  override def exitExpCast(ctx: BilParser.ExpCastContext): Unit = ctx.CAST().getText match {
    case "pad" => exprs.put(ctx, Pad(getExpr(ctx.exp), ctx.nat.getText.toInt))
    case "extend" => exprs.put(ctx, Extend(getExpr(ctx.exp), ctx.nat.getText.toInt))
    case "low" | "high" => exprs.put(ctx, getExpr(ctx.exp))
  }

  override def exitExpLoad(ctx: BilParser.ExpLoadContext): Unit =
    if (ctx.exp(0).getText == "mem") exprs.put(ctx, MemLoad(getExpr(ctx.exp(1)), Some(typeToSize(ctx.nat.getText))))
    else throw new AssumptionViolationException("Found load on variable other than mem")
    
  override def exitExpLoad8(ctx: BilParser.ExpLoad8Context): Unit =
    if (ctx.exp(0).getText == "mem") exprs.put(ctx, MemLoad(getExpr(ctx.exp(1)), Some(8)))
    else throw new AssumptionViolationException("Found load on variable other than mem")
    
  override def exitExpStore(ctx: BilParser.ExpStoreContext): Unit =
    if (ctx.exp(0).getText == "mem") {
      exprs.put(ctx, MemStore(getExpr(ctx.exp(1)), getExpr(ctx.exp(2)), Some(typeToSize(ctx.nat.getText))))
    } else throw new AssumptionViolationException("Found store on variable other than mem")
    
  override def exitExpStore8(ctx: BilParser.ExpStore8Context): Unit =
    if (ctx.exp(0).getText == "mem") exprs.put(ctx, MemStore(getExpr(ctx.exp(1)), getExpr(ctx.exp(2)), Some(8)))
    else throw new AssumptionViolationException("Found store on variable other than mem")
    
  override def exitExpFunctionCall(ctx: BilParser.ExpFunctionCallContext): Unit =
    exprs.put(ctx, FunctionCall("old", ctx.argList.exp.asScala.map(a => getExpr(a)).toList))

  override def exitPredBinOp(ctx: PredBinOpContext): Unit =
    preds.put(ctx, new PredBinOp(ctx.predBop.getText, getPred(ctx.pred(0)), getPred(ctx.pred(1))))
    
  override def exitPredUniOp(ctx: PredUniOpContext): Unit = 
    preds.put(ctx, PredUniOp(ctx.uop.getText, getPred(ctx.pred)))
    
  override def exitPredBracket(ctx: PredBracketContext): Unit = preds.put(ctx, getPred(ctx.pred))
  
  override def exitPredExprComp(ctx: PredExprCompContext): Unit =
    preds.put(ctx, ExprComp(ctx.expComp.getText, getExpr(ctx.exp(0)), getExpr(ctx.exp(1))))
  /* override def exitPredLiteral(ctx: BilParser.PredLiteralContext): Unit = preds.put(ctx, ctx.getText match {
    case "TRUE" => Bool.True
    case "FALSE" => Bool.False
  }) */

  override def exitSecLatticeElem(ctx: SecLatticeElemContext): Unit = secs.put(ctx, SecVar(ctx.getText))

  override def exitGamma(ctx: GammaContext): Unit = getExpr(ctx.`var`) match {
    case v: Register => gammaMappings.put(v, SecVar(ctx.ID.getText))
  }

  override def exitLpred(ctx: BilParser.LpredContext): Unit = (getExpr(ctx.`var`), getSec(ctx.secExpr)) match {
    case (v: Register, p: Sec) => lPreds.put(v, p)
  }

  override def exitRely(ctx: BilParser.RelyContext): Unit = rely = Some(getPred(ctx.pred))
  
  private def uniquePc() = {
    pcCount += 1
    "l" + pcCount
  }
}
